src/uucore/src/lib/features/checksum.rs

```diff
@@ -2,26 +2,967 @@
+                Ok(f) => Box::new(f),
+                Err(_) => {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        format!(
+                            "{}: No such file or directory",
+                            filename_input.to_string_lossy()
+                        ),
+                    )
+                    .into());
+                }
+            }
+        };
+
+        let reader = BufReader::new(file);
+        let lines: Vec<String> = reader.lines().collect::<Result<_, _>>()?;
+        let (chosen_regex, algo_based_format) =
+            determine_regex(filename_input, input_is_stdin, &lines)?;
+
+        // Process each line
+        for (i, line) in lines.iter().enumerate() {
+            if let Some(caps) = chosen_regex.captures(line) {
+                properly_formatted = true;
+
+                let mut filename_to_check = caps.name("filename").unwrap().as_str();
+                if filename_to_check.starts_with('*')
+                    && i == 0
+                    && chosen_regex.as_str() == SINGLE_SPACE_REGEX
+                {
+                    // Remove the leading asterisk if present - only for the first line
+                    filename_to_check = &filename_to_check[1..];
+                }
+
+                let expected_checksum = caps.name("checksum").unwrap().as_str();
+
+                // If the algo_name is provided, we use it, otherwise we try to detect it
+                let (algo_name, length) = if algo_based_format {
+                    // When the algo-based format is matched, extract details from regex captures
+                    let algorithm = caps.name("algo").map_or("", |m| m.as_str()).to_lowercase();
+
+                    // check if we are called with XXXsum (example: md5sum) but we detected a different algo parsing the file
+                    // (for example SHA1 (f) = d...)
+                    // Also handle the case cksum -s sm3 but the file contains other formats
+                    if algo_name_input.is_some() && algo_name_input != Some(&algorithm) {
+                        bad_format += 1;
+                        properly_formatted = false;
+                        continue;
+                    }
+
+                    if !SUPPORTED_ALGORITHMS.contains(&algorithm.as_str()) {
+                        // Not supported algo, leave early
+                        properly_formatted = false;
+                        continue;
+                    }
+
+                    let bits = caps.name("bits").map_or(Some(None), |m| {
+                        let bits_value = m.as_str().parse::<usize>().unwrap();
+                        if bits_value % 8 == 0 {
+                            Some(Some(bits_value / 8))
+                        } else {
+                            properly_formatted = false;
+                            None // Return None to signal a parsing or divisibility issue
```

Comment:
I think in case of a parse issue you don't reach this code.
```suggestion
                            None // Return None to signal a divisibility issue
```